File: backend/app/schemas/moderation.py
from datetime import datetime
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class ModerationContentType(str, Enum):
    PRODUCT = "Product"
    REVIEW = "Review"
    ARTISAN = "Artisan"
    STORY = "Story"


class ModerationStatus(str, Enum):
    PENDING = "Pending"
    APPROVED = "Approved"
    REJECTED = "Rejected"


class ModerationPriority(str, Enum):
    NORMAL = "Normal"
    HIGH = "High"


class ModerationCreate(BaseModel):
    content_type: ModerationContentType
    content_id: str = Field(min_length=1, max_length=100)
    submitted_by: UUID | None = None
    title: str = Field(min_length=1, max_length=255)
    description: str | None = None
    image_url: str | None = None
    priority: ModerationPriority = ModerationPriority.NORMAL


class ModerationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    content_type: ModerationContentType
    content_id: str
    submitted_by: UUID | None
    title: str
    description: str | None
    image_url: str | None
    status: ModerationStatus
    priority: ModerationPriority
    reviewed_by: UUID | None
    reviewed_at: datetime | None
    rejection_reason: str | None
    created_at: datetime
    updated_at: datetime


class ModerationReject(BaseModel):
    reason: str = Field(min_length=1, max_length=1000)

File: frontend/lib/api/moderation.ts
import api from "./client";

export type ModerationContentType =
  | "Product"
  | "Review"
  | "Artisan"
  | "Story";

export type ModerationStatus =
  | "Pending"
  | "Approved"
  | "Rejected";

export type ModerationPriority =
  | "Normal"
  | "High";

export type ModerationItem = {
  id: number;
  content_type: ModerationContentType;
  content_id: string;
  submitted_by?: string | null;
  title: string;
  description?: string | null;
  image_url?: string | null;
  status: ModerationStatus;
  priority: ModerationPriority;
  reviewed_by?: string | null;
  reviewed_at?: string | null;
  rejection_reason?: string | null;
  created_at: string;
  updated_at: string;
};

export type ModerationListParams = {
  status?: ModerationStatus;
  content_type?: ModerationContentType;
  priority?: ModerationPriority;
};

export type ModerationCreateData = {
  content_type: ModerationContentType;
  content_id: string;
  submitted_by?: string | null;
  title: string;
  description?: string | null;
  image_url?: string | null;
  priority?: ModerationPriority;
};

export type ModerationRejectData = {
  reason: string;
};

export const moderationApi = {
  list: (params?: ModerationListParams) =>
    api.get<ModerationItem[]>("/admin/moderation", {
      params,
    }),

  getById: (id: number) =>
    api.get<ModerationItem>(`/admin/moderation/${id}`),

  create: (data: ModerationCreateData) =>
    api.post<ModerationItem>("/admin/moderation", data),

  approve: (id: number) =>
    api.patch<ModerationItem>(
      `/admin/moderation/${id}/approve`,
    ),

  reject: (id: number, data: ModerationRejectData) =>
    api.patch<ModerationItem>(
      `/admin/moderation/${id}/reject`,
      data,
    ),
};
